src/v/redpanda/application.cc

```diff
@@ -870,6 +877,13 @@ void application::wire_up_redpanda_services() {
           std::ref(archival_scheduler),
           make_upload_controller_config(_scheduling_groups.archival_upload()))
           .get();
+
+        // In order to stop segment uploads and downloads we need
+        // to stop cloud_storage_api before archival_scheduler,
+        // recovery_manager, and other services that can upload or download
+        // objects from S3. Because of that it has to finish initialization
+        // later than all objects that depend on it.
+        finish_constructing_service(cloud_storage_api);
```

Comment:
I have a question, as `ss::sharded<>::stop()` destroys all underlying service instances is that guaranteed that after being stopped no other service will access `cloud_storage_api` in any way ? Maybe it would be better do decouple stopping from destroying the service instances ?

Reply:
it was working but probably just because of the coincidence 
I reworked this a bit